### PR TITLE
layers: Rework shader variable access logic

### PR DESCRIPTION
**Pull request by spencer-lunarg**
the `is_written_to`/`is_read_from` was a lie and never worked

This adds proper support to know if the variable is actually accessed. This is important for many upcoming SPIR-V related changes

### CODE REVIEW

File: layers/drawdispatch/descriptor_validator.cpp [hunk 1]
```diff
@@ -531,19 +526,16 @@ bool vvl::DescriptorValidator::ValidateDescriptor(const DescriptorBindingInfo &b
...
-                    continue;
-                }
+                if (!stage.entrypoint) continue;
                 for (const auto &inteface_variable : stage.entrypoint->resource_interface_variables) {
                     if (inteface_variable.decorations.set == set_index && inteface_variable.decorations.binding == binding) {
-                        descriptor_written_to |= inteface_variable.is_written_to;
-                        descriptor_read_from |=
-                            inteface_variable.is_read_from | inteface_variable.info.is_sampler_implicitLod_dref_proj;
-                        break;
+                        descriptor_written_to |= inteface_variable.IsWrittenTo();
```
Comment (filnet) [inline layers/drawdispatch/descriptor_validator.cpp:536]: typo in `inteface`

File: layers/state_tracker/shader_module.cpp [hunk 2]
```diff
@@ -401,6 +401,47 @@ static void FindPointersAndObjects(const Instruction& insn, vvl::unordered_set<u
...
 
+// Built-in can be both on the OpVariable or a inside a OpTypeStruct for Block built-in.
+bool EntryPoint::IsBuiltInWritten(spv::BuiltIn built_in, const Module& module_state, const StageInteraceVariable& variable,
+                                  const AccessChainVariableMap& access_chain_map) {
+    if (!variable.IsWrittenTo()) {
+        return false;
+    }
+    if (built_in == variable.decorations.builtin) {
+        return true;  // The built-in is on the Variable
+    } else if (!variable.type_struct_info || variable.type_struct_info->decorations.member_decorations.empty()) {
```
Comment (filnet) [inline layers/state_tracker/shader_module.cpp:412]: no need for else

File: layers/state_tracker/shader_module.h [hunk 3]
```diff
@@ -373,15 +408,13 @@ struct ResourceInterfaceVariable : public VariableBase {
         bool is_zero_extended{false};  // if at least one access has ZeroExtended
 
         // If a variable is used as a function arguement, but never actually used, it will be found in EntryPoint::accessible_ids so
-        // we need to have a dedicated mark if it was accessed
-        // TODO - merge with is_read_from and is_written_to
-        bool is_image_accessed{false};
+        // we need to have a dedicated mark if it was accessed.
+        // We use this for variable hasing, but the VariableBase has the helper functions to read this value.
```
Comment (filnet) [inline layers/state_tracker/shader_module.h:412]: hasing => hashing ?

File: layers/state_tracker/shader_module.h [hunk 4]
```diff
@@ -301,7 +336,8 @@ struct StageInteraceVariable : public VariableBase {
     const std::vector<uint32_t> builtin_block;
     uint32_t total_builtin_components = 0;
 
-    StageInteraceVariable(const Module &module_state, const Instruction &insn, VkShaderStageFlagBits stage);
+    StageInteraceVariable(const Module &module_state, const Instruction &insn, VkShaderStageFlagBits stage,
```
Comment (filnet) [inline layers/state_tracker/shader_module.h:339]: Interace ?

Comment (spencer-lunarg) [reply layers/state_tracker/shader_module.h]: this was is particularly embarrassing... at this rate, I might need to add spellcheck CI for myself
Thanks for catching these

File: layers/state_tracker/shader_module.h [hunk 5]
```diff
@@ -186,6 +186,34 @@ struct TypeStructInfo {
     TypeStructSize GetSize(const Module &module_state) const;
 };
 
+enum AccessBit {
+    empty = 0,
```
Comment (artem-lunarg) [inline layers/state_tracker/shader_module.h:190]: It looks like enum names (emtpy, read, write) go directly into the global namespace. Can they be localized? struct or namespace AccessBit with constants or nested enum, or better alternative.

Comment (spencer-lunarg) [reply layers/state_tracker/shader_module.h:190]: I think the namespace and constants work well here

Comment (artem-lunarg) [reply layers/state_tracker/shader_module.h]: Ah, that's updated now...

**Review by artem-lunarg — APPROVED** [review]
LGTM. To be honest don't understand all details but checked general structure and trust the expertise of the author.

**Pull request merged**